README
Text-SenseClusters-Wikipedia-LabelEvaluation

Version 0.01

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.



INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install


SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Text::SenseClusters::Wikipedia::LabelEvaluation

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Text-SenseClusters-Wikipedia-LabelEvaluation

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Text-SenseClusters-Wikipedia-LabelEvaluation

    CPAN Ratings
        http://cpanratings.perl.org/d/Text-SenseClusters-Wikipedia-LabelEvaluation

    Search CPAN
        http://search.cpan.org/dist/Text-SenseClusters-Wikipedia-LabelEvaluation/


LICENSE AND COPYRIGHT

Copyright (C) 2012 Ted Pedersen, Anand Jha

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for Text-SenseClusters-Wikipedia-LabelEvaluation

0.05    Sun Dec 2 13:32:41 CST 2012
        Fifth version, added the various input parameters of the program as hash instead of taking all the parameters individually.


0.04    Sun Nov 18 19:28:13 CST 2012
        Fourth version, restructure label evaluation program to have modular structures and added an example of using this module in SYNOPSIS.




lib/Text/SenseClusters/Wikipedia/GetWikiData.pm
#!/usr/bin/perl -w

# Declaring the Package for the module.
package Text::SenseClusters::Wikipedia::GetWikiData;

use strict; 
use encoding "utf-8";

# The following two lines will make this module inherit from the Exporter Class.
require Exporter;
our @ISA = qw(Exporter);

# Using WWW::Wikipedia Module.
# Reference: http://search.cpan.org/dist/WWW-Wikipedia/lib/WWW/Wikipedia.pm
use WWW::Wikipedia;

# Defining the Variable for using the Wikipedia Module.
# Reference: http://search.cpan.org/~bricas/WWW-Wikipedia-2.00/
my $wiki = WWW::Wikipedia->new();


##########################################################################################
=head1 function: getWikiDataForTopic
---------------------------------------------------

	This function will fetch data about a topics from the Wikipedia, then it  
	will write the fetched data into a new file 'topic_Name.txt'.


	@argument1	: Name of the topic for which we need to fetch data from the 
 				  Wikipedia.

	@return 	: Name of the file in which this function has written the 
				  data,'topic_Name.txt'.
		  
	@description	:
		1). Reading the topic to read from the function arguments.
		2). Use this topic name to create file name in which we will write
			data about the topic.
 		3). Get the data from the Wikipedia module about the topic and write
			it into the above mentioned topic.
		4). Return the file name.

=cut
##########################################################################################
sub getWikiDataForTopic{
	
	# Read the Topic name from the argument of the function.
	my $topicToLook = shift;
	
	# Removing the white space from the front and end of the word.
	$topicToLook =~ s/^\s+|\s+$//g;

	# Removing the white space with underscore.
	$topicToLook =~ s/\s+/_/g;

	# Creating the fileName from the topic name.
	my $fileName = "temp_$topicToLook.txt";
	
	# Open the file handle in Write Mode.
	open (MYFILE, ">$fileName");

	# Use Wikipedia Search to get the result about the topic.
	# Reference: http://search.cpan.org/~bricas/WWW-Wikipedia-2.00/	
	my $result = $wiki->search($topicToLook);

	# If the entry has some text, write it out to file.
	if ($result){
		# Writing the content of the search result into the newly created file.  
		print MYFILE $result->text();
		
		# Also writing the list of any related items into the files. 
		print MYFILE join( "\n", $result->related() );
	}

	# Close the file handle.
	close (MYFILE);

	# Returning the name of the file in which we write the Wikipedia data 
	# about the given topic.
	return $fileName;
}



lib/Text/SenseClusters/Wikipedia/SimilarityScore.pm
#!/usr/bin/perl -w

# Declaring the Package for the module.
package Text::SenseClusters::Wikipedia::SimilarityScore;

use strict; 
use encoding "utf-8";

# The following two lines will make this module inherit from the Exporter Class.
require Exporter;
our @ISA = qw(Exporter);

# Using WWW::Wikipedia Module.
# Reference: http://search.cpan.org/dist/WWW-Wikipedia/lib/WWW/Wikipedia.pm
use WWW::Wikipedia;

# Defining the Variable for using the Wikipedia Module.
# Reference: http://search.cpan.org/~bricas/WWW-Wikipedia-2.00/
my $wiki = WWW::Wikipedia->new();

# Using Text Similarity Module.
# Reference: http://search.cpan.org/~tpederse
#					/Text-Similarity-0.08/lib/Text/Similarity.pm
use Text::Similarity::Overlaps;


########################################################################################
=head1 Function: computeOverlappingScores
------------------------------------------------


	Function that will compare the labels file with the wiki files and  
	will return the overlapping score. 

	@argument1		: Name of the cluster file.
	@argument2		: Name of the file containing the data from Wikipedia.
	@argument3		: Name of the file containing the stop word lists.
 
	@return 		: Return the overlapping score between these files.
		  
	@description	:
		1). Reading the file name from the command line argument.
		2). Invoking the Text::Similarity::Overlaps module and passing
			the file names for similarity comparison.
 		3). Then overlapping score obtained from this module is returned 
			as the similarity value.

=cut

#########################################################################################

sub computeOverlappingScores{
	 
	# Getting the ClusterFileName from the argument.
	my $clusterFileName = shift;
	
	# Getting the TopicFileName from the argument.
	my $topicFileName = shift;

	# Getting the stop list file location.
	my $stopListFileLocation = shift;
	
	if(!defined $stopListFileLocation){
			 # Getting the module name.
			my $module = "Text/SenseClusters/Wikipedia/SimilarityScore.pm";
			   
			# Finding its installed location.
			my $moduleInstalledLocation = $INC{$module};
		
			# Getting the prefix of installed location. This will be one of 
			# the values in array @INC.
			$moduleInstalledLocation =~ 
				m/(.*)Text\/SenseClusters\/Wikipedia\/SimilarityScore\.pm$/g;
			
			# Getting the installed stopList.txt location using above location. 
			# For e.g.:
			#	/usr/local/share/perl/5.10.1/Text/SenseClusters
			#			/Wikipedia/stoplist.txt
			$stopListFileLocation 
					= $1."/Text/SenseClusters/Wikipedia/stoplist.txt";
			

	}
	# Setting the Options for getting the results from the Text::Similarity
	# Module.
	my %options = ('verbose' => 0, 'stoplist' => $stopListFileLocation);

	# Creating the new Overlaps Object.
	my $mod = Text::Similarity::Overlaps->new (\%options);
	
	# If the object is not created, then quit the program with error message. 
	defined $mod or die "Construction of Text::Similarity::Overlaps failed";

	# Getting the overlapping score from the Similarity function.
	my $score = $mod->getSimilarity ($clusterFileName, $topicFileName);

	# Printing the Similarity Score for the files.
	# print "The similarity of $clusterFile and $topicFile is : $score\n";
	
	# Returning the overlapping Score.
	return $score;
}




lib/Text/SenseClusters/Wikipedia/PrintingHashData.pm
#!/usr/bin/perl -w

# Declaring the Package for the module.
package Text::SenseClusters::Wikipedia::PrintingHashData;

use strict; 
use encoding "utf-8";

# The following two lines will make this module inherit from the Exporter Class.
require Exporter;
our @ISA = qw(Exporter);


##########################################################################################
=head1 Function: prinHashOfHash
------------------------------------------------

	Function to print the content of  Hash-of-Hash.
 
	@argument1	: Reference of HashOfHash whose values has to be printed. 
	@return 		: Nothing.
				  
	Description:
			1. Iterate through the outer key in sorted order.
			2. Iterate through the inner key and print the corresponding value.

=cut
##########################################################################################
sub prinHashOfHash{
	# Getting the Hash Reference from the argument.
	my $hashOfHashRef = shift; 
	
	# Getting the Hash from its reference.
	my %hashOfHash = %$hashOfHashRef;

	# Step 1: Iterating through the OuterKeys Of the Hash.
	foreach my $sortedOuterKey (sort keys %hashOfHash){
		print "\n\nOuterKey = '$sortedOuterKey'";
	   
	    # Step 2: Iterating through the InnerKeys Of the Hash.
	    foreach my $sortedInnerKey (sort keys %{$hashOfHash{$sortedOuterKey}}){

			# Step 3:Printing the value.
	        print "\nInnerKey=$sortedInnerKey  ".
	        		"Value=$hashOfHash{$sortedOuterKey}{$sortedInnerKey}";        
	    }    
	}
}


##########################################################################################
=head1 Function: prinHashOfScore
------------------------------------------------

	This function will print the score of each cluster and its most
	probable against a topic and its corresponding score.



	For e.g:
	Direct Col Conclusion::	
		Cluster0 		:	Tony_Blair 		,	 0.577
		Cluster1 		:	Bill_Clinton 		,	 0.571
	Direct Row Conclusion::	
		Bill_Clinton 		:	Cluster1	 	,	 0.522
		Tony_Blair 		:	Cluster0 		,	 0.625

	Inverse Row Conclusion::	
		Cluster0 		:	Tony_Blair 		,	 0.625
		Cluster1 		:	Bill_Clinton 		,	 0.522
	Inverse Col Conclusion::	
		Bill_Clinton 		:	Cluster1 		,	 0.571
		Tony_Blair 		:	Cluster0 		,	 0.577

	@Argument	: Reference of HashOfHash 
				 (i)  containing the topic (with supporting score) against a 
				 	   cluster name.
				 (ii) containing the cluster name (with supporting score) 
				 	   against a topic.

	@return 	: Nothing.

	@Description:
			1. Get the HashOfHash Reference from the function argument.
			2. Iterate through key in sorted order.
			3. Clean the key and value.
			4. Write the result into the Output file.

	Output File:

	    Output file is the final file that user can see to get the detailed 
		result about the complete evaluation process.

=cut 
##########################################################################################

sub prinHashOfScore{
	# Getting the Hash Reference from the argument.
	my $topicClusterHashRef = shift; 
	
	# Getting the Hash from its reference.
	my %topicClusterHash = %$topicClusterHashRef;

	# Getting the File Handle from the function argument.
	my $outputFileHandle = shift;

	# Going through Hash and Printing the score.
	# Getting the Key of the Hash in the sorted order.
	foreach my $sortedKey (sort keys %topicClusterHash){
		
		# Extracting the "topic" from the key (FileName is the key).
		my $key = $sortedKey; 	
		
		# Removing the additional text from the key (filename related stuff).
		$key =~ s/temp_//;
		$key =~ s/.txt//;
		
		# Extracting the value from the hash.
		my $value = $topicClusterHash{$sortedKey};
		
		# Removing the additional text from the values (filename related stuff).
		$value =~ s/temp_//;
		$value =~ s/.txt//;
		
		# Writing the data into the output file.
		print $outputFileHandle "\n \t$key \t:\t$value"; 
	}
}



lib/Text/SenseClusters/Wikipedia/ReadingFilesData.pm
#!/usr/bin/perl -w

package Text::SenseClusters::Wikipedia::ReadingFilesData;

use strict; 
use encoding "utf-8";

# The following two lines will make this module inherit from the Exporter Class.
require Exporter;
our @ISA = qw(Exporter);

###########################################################################################

=head1 Function: readLinesFromClusterFile
------------------------------------------------

	This function will read lines from the file containing the Labels of the  
	Clusters and make the hash file.

	@argument1	: Name of the cluster file name.
	@argument2	: Reference of Hash ($labelSenseClustersHash) which will hold   
				  the information in the following format:
 					For e.g.: Cluster0{
									Descriptive    => George Bush, Al Gore, White 
														House, New York
									Discriminating => George Bush, York Times	
 							  } 
							  Cluster1{
									Descriptive    => George Bush, BRITAIN London, 
														 Prime Minister
									Discriminating => BRITAIN London, Prime Minister	
 							  } 

	@return 	: It will return the reference of the Hash mentioned above: 
		  			$labelSenseClustersHashRef.
		  			
	@description	:
		1. Read the file line by line.
		
 		2. Ignore the lines which do not follow one of the following format:
     		 Cluster 0 (Descriptive): George Bush, Al Gore, White House, New York
     		 Cluster 0 (Discriminating): George Bush, BRITAIN London
     		 
 		3. Create Key from the "Cluster # (Descriptive)" or "Cluster # (Discrim
			- inating)" as "OuterKey: Cluster#" "InnerKey: Descriptive".
			
 		4. Store the value of hash as the keywords similar to above example:
			for e.g:
			
			  $labelSenseClustersGlobalRef{Cluster0}{Discriminating}
			  			= "BRITAIN London, Prime Minister";
			  
			  
=cut
###########################################################################################

sub readLinesFromClusterFile{ 

	# Reading the cluster file Name from the argument.
	my $clusterFileName = shift;
	
	# Reading the reference from the argument.
	my $labelSenseClustersHashRef = shift;
		
	# Getting the hash from the reference.	
	my %labelSenseClustersHash = %$labelSenseClustersHashRef;

	# Opening the File passed by user as the first command line argument.
	# It should be the name of the cluster file containing the labels.
	open clusterFile, $clusterFileName or die $!;

	while (<clusterFile>){
		# Removing the new line character.
		chomp;
		
		# Removing the white space from the front and end of the word.
		$_ =~ s/^\s+|\s+$//g;
		
		# If the line is empty then ignore that line and go to next line.	
		if($_ eq ''){
			next;
		}

		# Contents of LabelFile.
		#     Cluster 0 (Descriptive): George Bush, Al Gore, White House, New York
		#     Cluster 0 (Discriminating): George Bush, BRITAIN London
		
		# Spiliting each line by ":". 
		my @lineArray = split(/:/, $_);
		
		# If the given do not have Two elements after split. (It means no data for the 
		# given cluster.) Then ignore that cluster.
		if(scalar(@lineArray)!=2){
		  next;
		}

		# Following Code are for making the Key (which will be Cluster Number and Type of 
		# Labels) Typical Key Structure --> "Cluster 0 (Descriptive)"
		
		# Spiliting the elements contianing the information about the key with whitespace
		my @keyArray = split(/\s+/, $lineArray[0]);	
		
		# If something wrong with the structure than ignore the key and carry on with 
		# next line.
		if(scalar(@keyArray)!=3){
			next;
		}
		
		# Making of the Outer key, which is "cluster#"
		my $outerKey = $keyArray[0].$keyArray[1];
		
		# The inner key indicates the type of label i.e. Descriptive or Discriminating. 
		my $innerKey = $keyArray[2];
		
		# Removing the start parenthesis '(' and closing ')' parenthesis from the inner 
		# key.
		$innerKey =~s/[(,)]+//g;
		
		# Setting the keywords associated with this keys as the value.
		# For e.g.: Cluster0{
		#				Descriptive =>		George Bush, Al Gore, White House, New York
		#				Discriminating =>	George Bush, BRITAIN London	
		# 			} 
		$labelSenseClustersHash{$outerKey}{$innerKey} = $lineArray[1];
	}
	
	# Close the file handle.
	close (clusterFile);  
	
	# Returning the reference of the Hash containg the Labels information from 
	# the cluster.
	return \%labelSenseClustersHash;
}



##########################################################################################
=head1 Function: readLinesFromTopicFile

------------------------------------------------

	This function will read lines from the topic file and list of all the 
	topics. 

	@argument1	: Name of the topicFile.
  
	@return 	: String containing the list of all the topics(labels) for  
				  the clusters.
	  
	@description	:
		1. Read the file line by line.
 		2. Remove the new line characters and making string variable which 
  		   contains the list of all the topics.
  		   
=cut
##########################################################################################

sub readLinesFromTopicFile{ 
	
	# Getting the topic file name from argument.
	my $topicFileName = shift;

	# Opening the File, whose name is passed as the second command-line-argument.
	# It is the name of the file which contains the list of the topics for clusters.
	open topicFile, $topicFileName or die $!;
	
	# Defining the variable which will hold all the topics.
	my $topicData = "";
	
	# Reading the file line by line till end of file.
	while (<topicFile>){
		
		# Removing the new line character.
		chomp;
		
		# Concatenating it to previous line.
		$topicData = $topicData.$_;
	}
	
	# Close the file handle.
	close (topicFile);  
		
	# Returning the topic list.
	return $topicData;
}

# Making the default return statement as 1;
# Reference : http://lists.netisland.net/archives/phlpm/phlpm-2001/msg00426.html

1;


Makefile.PL
use 5.006;
use strict;
use warnings;
use ExtUtils::MakeMaker;



# Reference: http://search.cpan.org/~mschwern/ExtUtils-MakeMaker-6.62/lib/ExtUtils/MakeMaker.pm
# Things added here:
#    NAME              = Name of the main module, which act as point of start.
#    AUTHOR            = Name of the author.
#    VERSION_FROM      = Place from where to find the version of the module.
#    ABSTRACT_FROM     = Place from where to find the abstract information about the module.
#    PREREQ_PM         = {}, # e.g., Module::Name => 1.1
#    EXE_FILES         = [map "bin/$_", @programs],


WriteMakefile(
    NAME                => 'Text::SenseClusters::Wikipedia::LabelEvaluation',
    AUTHOR              => q{Ted Pedersen <tpederse at d.umn.edu>},
    VERSION_FROM        => 'lib/Text/SenseClusters/Wikipedia/LabelEvaluation.pm',
    ABSTRACT_FROM       => 'lib/Text/SenseClusters/Wikipedia/LabelEvaluation.pm',
    ($ExtUtils::MakeMaker::VERSION ge '6.30_00'
      ? ('LICENSE'=> 'gpl')
      : ()),
    PL_FILES            => {},
    PREREQ_PM => {
        'Test::More' 		=> 0,
	'Text::Similarity' 	=> 0.08,
	'WWW::Wikipedia' 	=> 2.00
    },
    dist                => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean               => { FILES => 'Text-SenseClusters-Wikipedia-LabelEvaluation-*' },
);
